spec.html

```diff
@@ -37627,29 +37597,55 @@ <h1>Array.prototype.findLast ( _predicate_ [ , _thisArg_ ] )</h1>
-            1. Let _Pk_ be ! ToString(𝔽(_k_)).
-            1. Let _kValue_ be ? Get(_O_, _Pk_).
-            1. Let _testResult_ be ToBoolean(? Call(_predicate_, _thisArg_, « _kValue_, 𝔽(_k_), _O_ »)).
-            1. If _testResult_ is *true*, return 𝔽(_k_).
-            1. Set _k_ to _k_ - 1.
-          1. Return *-1*<sub>𝔽</sub>.
+          1. Let _findRec_ be ? FindViaPredicate(_O_, _len_, ~descending~, _predicate_, _thisArg_).
+          1. Return _findRec_.[[Index]].
         </emu-alg>
         <emu-note>
           <p>This method is intentionally generic; it does not require that its *this* value be an Array object. Therefore it can be transferred to other kinds of objects for use as a method.</p>
         </emu-note>
+
+        <emu-clause id="sec-findviapredicate" type="abstract operation">
+          <h1>
+            FindViaPredicate (
+              _O_: an Object,
+              _len_: a non-negative integer,
+              _direction_: ~ascending~ or ~descending~,
+              _predicate_: an ECMAScript language value,
+              _thisArg_: an ECMAScript language value,
+            ): either a normal completion containing a Record with fields [[Index]] (an integral Number) and [[Value]] (an ECMAScript language value) or a throw completion
+          </h1>
+          <dl class="header">
+            <dt>description</dt>
+            <dd>
+              <p>_O_ should be an array-like object or a TypedArray. This operation calls _predicate_ once for each element of _O_, in either ascending index order or descending index order (as indicated by _direction_), until it finds one where _predicate_ returns a value that coerces to *true*. At that point, this operation returns a Record that gives the index and value of the element found. If no such element is found, this operation returns a Record that specifies *-1*<sub>𝔽</sub> for the index and *undefined* for the value.</p>
+              <p>_predicate_ should be a function. When called for an element of the array, it is passed three arguments: the value of the element, the index of the element, and the object being traversed. Its return value will be coerced to a Boolean value.</p>
+              <p>_thisArg_ will be used as the *this* value for each invocation of _predicate_.</p>
+              <p>This operation does not directly mutate the object on which it is called, but the object may be mutated by the calls to _predicate_.</p>
+              <p>The range of elements processed is set before the first call to _predicate_, just before the traversal begins. Elements that are appended to the array after this will not be visited by _predicate_. If existing elements of the array are changed, their value as passed to _predicate_ will be the value at the time that this operation visits them. Elements that are deleted after traversal begins and before being visited are still visited and are either looked up from the prototype or are *undefined*.</p>
```

Comment:
```suggestion
              <p>This operation does not directly mutate the object on which it is called, but the object may be mutated by the calls to _predicate_. If existing elements of the array are changed, their value as passed to _predicate_ will be the value at the time that this operation visits them. Elements that are deleted after traversal begins and before being visited are still visited and are either looked up from the prototype or are *undefined*.</p>
```

Reply:
To be clear, the suggestion deletes the two sentences:
> The range of elements processed is set before the first call to _predicate_, just before the traversal begins. Elements that are appended to the array after this will not be visited by _predicate_. 

Personally, I think those sentences are worth keeping.

Reply:
I agree with jmdyck that the existing sentences are clearer and should be kept. Specifically they directly describe the subtlety around snapshotting, while I find the length-arithmetic sentence to be indirectly imply the snapshotting semantics. Let's keep the existing two sentences.